cclib/method/ddec.py

```diff
@@ -631,67 +685,258 @@ def _calculate_w_and_u(self):
+                        )
+
+                        # Update upperphi and lowerphi
+                        lower_ind = numpy.where(
+                            self._candidates_phi[atomi]
+                            == self._candidates_phi[atomi][self._candidates_phi[atomi] < 0].max()
+                        )[0][0]
+                        upper_ind = numpy.where(
+                            self._candidates_phi[atomi]
+                            == self._candidates_phi[atomi][self._candidates_phi[atomi] > 0].min()
+                        )[0][0]
+
+                        lowerphi = self._candidates_phi[atomi][lower_ind]
+                        upperphi = self._candidates_phi[atomi][upper_ind]
+
+                        if abs(lowerphi) < self.convergence_level:
+                            bigphiAII[atomi] = self._candidates_bigPhi[atomi][lower_ind]
+                            phiconverged = True
+                        elif abs(upperphi) < self.convergence_level:
+                            bigphiAII[atomi] = self._candidates_bigPhi[atomi][upper_ind]
+                            phiconverged = True
+                        else:
+                            # Fitting needs to continue in this case.
+                            lowerbigPhi = self._candidates_bigPhi[atomi][lower_ind]
+                            lowerphi = self._candidates_phi[atomi][lower_ind]
+                            upperbigPhi = self._candidates_bigPhi[atomi][upper_ind]
+                            upperphi = self._candidates_phi[atomi][upper_ind]
+
+            assert phiconverged, "Iterative conditioning failed to converge."
+
+            # Set final G_A value using chosen Phi
+            self._g[atomi] = self._phiaii(self.rho_wavg[atomi], eta, bigphiAII[atomi], atomi)[1]
+
+    def calculate_H(self):
+        """ Calculate H_A(r_A)
+            [STEP 4-7]
+        """
+        self._h = []
+        for atomi in range(len(self._g)):
+            # First set H_est as G_A
+            self._h.append(self._g[atomi])
+
+            # Determine eta_upper using equation 86 in doi: 10.1039/c6ra04656h
+            # and apply upper limit using equation 91.
+            eta = 2.5 * convertor(1, "Angstrom", "bohr") / (1 - (self.tau[atomi]) ** 2)
+            exp_applied = self._h[atomi][:-1] * numpy.exp(
+                -1 * eta[1:] * numpy.diff(self.radial_grid_r[atomi])
+            )
+            for radiusi in range(1, len(self._h[atomi])):
+                self._h[atomi][radiusi] = max(self._h[atomi][radiusi], exp_applied[radiusi - 1])
+
+            # Normalize using equation 92 in doi: 10.1039/c6ra04656h.
+            self._h[atomi] = (
+                self._h[atomi]
+                * self._integrate_from_radial([self._g[atomi]], [atomi])
+                / self._integrate_from_radial([self._h[atomi]], [atomi])
+            )
+
     def _phiai(self, ya, bigphiAI, atomi):
```

Comment:
Better method name?

Reply:
I've changed this (and _phiaii) into _update_phiai (and update_phiaii). Thanks.